src/Storages/MergeTree/MergeTreeSelectProcessor.cpp

```diff
@@ -256,9 +266,54 @@ void MergeTreeSelectProcessor::initializeReadersChain()
     task->initializeReadersChain(all_prewhere_actions, read_steps_performance_counters);
 }
 
-Block MergeTreeSelectProcessor::transformHeader(Block block, const PrewhereInfoPtr & prewhere_info)
+void MergeTreeSelectProcessor::injectLazilyReadColumns(
+    size_t rows,
+    Block & block,
+    MergeTreeReadTask * task,
+    const LazilyReadInfoPtr & lazily_read_info)
+{
+    if (!lazily_read_info)
+        return;
+
+    ColumnPtr row_num_column;
+    ColumnPtr part_num_column;
+    if (rows)
+    {
+        row_num_column = block.getByName("_part_offset").column;
+        part_num_column = DataTypeUInt64().createColumnConst(rows, task->getInfo().part_index_in_query)->convertToFullColumnIfConst();
```

Comment:
There is a concern about part_num column being a constant column, since chunks can be squashed in some pipelines. Currently, I don't see that this optimization can be applied where we use squashing. But anyway, we'll check this concern separately and address it if necessary in follow-up PR.